src/libxtp/gwbse/bse.cc

```diff
@@ -655,5 +660,93 @@ BSE::Interaction BSE::Analyze_eh_interaction(const QMStateType& type,
   return analysis;
 }
 
+void BSE::Perturbative_DynamicalScreening(
+    const QMStateType& type, const Orbitals& orb,
+    const Eigen::VectorXd& Hd_static_contrib) {
+
+  Interaction analysis;
+
+  const tools::EigenSystem& BSECoefs =
+      (type == QMStateType::Singlet) ? orb.BSESinglets() : orb.BSETriplets();
+
+  const Eigen::VectorXd& RPAInputEnergies = orb.RPAInputEnergies();
+  const Eigen::VectorXd& BSEenergies = BSECoefs.eigenvalues();
+
+  // initial copy of static BSE energies to dynamic
+  Eigen::VectorXd BSEenergies_dynamic = BSEenergies;
+
+  // recalculate Hd at the various energies
+  for (Index i_exc = 0; i_exc < BSEenergies.size(); i_exc++) {
+    XTP_LOG(Log::info, _log) << "Dynamical Screening BSE, Excitation " << i_exc
+                             << " static " << BSEenergies(i_exc) << flush;
+
+    for (Index iter = 0; iter < _max_dyn_iter; iter++) {
```

Comment:
Have you thought about using your anderson mixing class here as well to keep the number of iterations lower?

Reply:
Not really. From my, admittedly small test, but also previous experiences with the super old fortran code, this usually is done in 2-3 evaluations.